types/validator_set.go

```diff
@@ -368,63 +368,84 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
 	if !blockID.Equals(commit.BlockID) {
 		return fmt.Errorf("Invalid commit -- wrong block id: want %v got %v",
 			blockID, commit.BlockID)
 	}
 
-	talliedVotingPower := int64(0)
+	var talliedVotingPower int64
 	round := commit.Round()
 
 	for idx, precommit := range commit.Precommits {
+		// Some precommits will likely be missing, skip those
 		if precommit == nil {
-			continue // OK, some precommits can be missing.
+			continue
 		}
+
+		// Malicous data checking for height
 		if precommit.Height != height {
 			return fmt.Errorf("Invalid commit -- wrong height: want %v got %v", height, precommit.Height)
 		}
+
+		// Malicous data checking for round
 		if precommit.Round != round {
 			return fmt.Errorf("Invalid commit -- wrong round: want %v got %v", round, precommit.Round)
 		}
+
+		// Malicous data checking for precommit
 		if precommit.Type != PrecommitType {
 			return fmt.Errorf("Invalid commit -- not precommit @ index %v", idx)
 		}
-		_, val := vals.GetByIndex(idx)
+
 		// Validate signature.
-		precommitSignBytes := precommit.SignBytes(chainID)
-		if !val.PubKey.VerifyBytes(precommitSignBytes, precommit.Signature) {
+		_, val := vals.GetByAddress(precommit.ValidatorAddress)
+		if val == nil {
+			continue
+		}
+
+		// verify that the valiator signed the precommit
+		if !val.PubKey.VerifyBytes(precommit.SignBytes(chainID), precommit.Signature) {
 			return fmt.Errorf("Invalid commit -- invalid signature: %v", precommit)
 		}
+
 		// Good precommit!
 		if blockID.Equals(precommit.BlockID) {
 			talliedVotingPower += val.VotingPower
-		} else {
-			// It's OK that the BlockID doesn't match.  We include stray
-			// precommits to measure validator availability.
 		}
+
+		// It's OK that the BlockID doesn't match.  We include stray
+		// precommits to measure validator availability.
 	}
 
 	if talliedVotingPower > vals.TotalVotingPower()*2/3 {
 		return nil
 	}
+
 	return errTooMuchChange{talliedVotingPower, vals.TotalVotingPower()*2/3 + 1}
 }
 
-// VerifyFutureCommit will check to see if the set would be valid with a different
-// validator set.
+// VerifyFutureCommit checks to see if a given future validator set has
+// committed a block, and whether those who signed of this future validator set
+// has sufficient overlap with this validator set.
 //
-// vals is the old validator set that we know.  Over 2/3 of the power in old
-// signed this block.
+// vals is the current validator set that we know.  Over 2/3 of the power in
+// this valset is expected to have signed this block.
 //
-// In Tendermint, 1/3 of the voting power can halt or fork the chain, but 1/3
-// can't make arbitrary state transitions.  You still need > 2/3 Byzantine to
-// make arbitrary state transitions.
+// Justification for the 2/3: In Tendermint, 1/3 of the voting power can halt
+// or fork the chain, but 1/3 can't make arbitrary state transitions.  You
+// still need > 2/3 Byzantine to make arbitrary state transitions.
```

Comment:
Hmm, is this the right justification? I thought it was that we wanted the light client algorithm to guarantee that any set of commits which convinced a light client of a history that differed from the actual history of the canonical chain run by full nodes necessarily involved an equivocation, which would then be slashable on the main chain.

Also, I don't see what arbitrary state transitions have to do with bisection - 2/3 of voting power can sign anything even if we don't bisect at all, since the light client doesn't execute transactions.

Reply:
I think this might be reference to https://github.com/tendermint/tendermint/issues/3244#issuecomment-475921646 but I didn't quite understand that

Reply:
https://github.com/tendermint/tendermint/issues/3244#issuecomment-491366760

In any case, I am pretty sure we want the stronger property that any set of signatures which would succeed in fooling a light will necessarily cause a well-defined fraction of stake to be slashed - that economic cost is what's providing security to the light client, especially where lying may be particularly valuable (e.g. in IBC).